lib/TPP/Transforms/ConvertPackToExpandShape.cpp

```diff
@@ -0,0 +1,123 @@
+//===- FoldAddIntoDest.cpp ---------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Passes.h"
+#include "TPP/Transforms/Utils/ValueUtils.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Tensor/Utils/Utils.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/Interfaces/DestinationStyleOpInterface.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Casting.h"
+#include <cstdint>
+using namespace mlir;
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_CONVERTPACKTOEXPANDSHAPEPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+namespace {
+
+static FailureOr<std::pair<tensor::ExpandShapeOp, AffineMap>>
+packToExpandShape(tensor::PackOp packOp, linalg::GenericOp genericOp,
+                  PatternRewriter &rewriter) {
+  AffineMap affineMap;
+  // TODO: clean-up
+  for (auto &use : packOp->getUses()) {
+    affineMap = genericOp.getMatchingIndexingMap(&use);
+    break;
+  }
+  auto origShape =
+      dyn_cast<TensorType>(packOp->getOperand(0).getType()).getShape();
+  auto packedType = dyn_cast<TensorType>(packOp->getResult(0).getType());
+  auto packedShape = packedType.getShape();
+  auto packInverseMap = AffineMap::getPermutationMap(
+      mlir::tensor::getPackInverseDestPerm(packOp), rewriter.getContext());
+  auto normalizedShape = applyPermutationMap(packInverseMap, packedShape);
+
+  auto normalizedType = packedType.clone(normalizedShape);
+  auto normalizedIndexingMap = packInverseMap.compose(affineMap);
+
+  auto innerDimPos = SmallVector<unsigned int>(packOp.getInnerDimsPos());
+
+  SmallVector<SmallVector<int64_t, 2>> associationIndices;
+  int curDimIdx = 0;
+  for (auto idx : llvm::seq(origShape.size())) {
+    associationIndices.emplace_back(SmallVector<int64_t>());
+    associationIndices.back().push_back(curDimIdx++);
+    if (llvm::is_contained(innerDimPos, idx))
+      associationIndices.back().push_back(curDimIdx++);
+  }
+
+  auto expandShape = rewriter.create<tensor::ExpandShapeOp>(
+      genericOp.getLoc(), normalizedType, packOp.getOperand(0),
+      ArrayRef(associationIndices));
+  rewriter.replaceAllOpUsesWith(packOp, expandShape);
```

Comment:
I don't think we can update other users. It might be best to only update specific generic's operand.
Or maybe it doesn't make sense to rewrite pack at all if there are multiple users. But then if these are multiple generic contractions then maybe it's still fine.

For us, restriction to only one user might be enough, in general I'd leave that for a user to decide.

Reply:
Have now added requirement for single user for there to be a match.

We can choose to relax later.